Sources/SpelledPitch/PitchSpeller/Wetherfield/WeightedGraphSchemeProtocol.swift

```diff
@@ -21,25 +23,84 @@ extension WeightedGraphSchemeProtocol {
         H: WeightedGraphSchemeProtocol,
         H.Weight == Weight
     {
-        return H.init { self.weight(Edge(f($0.a),f($0.b))) }
+        return H { self.weight(from: f($0.a), to: f($0.b)) }
     }
-}
 
-extension WeightedGraphSchemeProtocol {
     @inlinable
     func unweighted <H> () -> H where
         H: UnweightedGraphSchemeProtocol,
         H.Edge == Edge
     {
-        return H.init { self.weight($0) != nil }
+        return H { self.weight($0) != nil }
+    }
+}
+
+extension WeightedGraphSchemeProtocol where Self: UndirectedGraphSchemeProtocol, Weight: Numeric {
+    
+    static func * (lhs: Self, rhs: Self) -> Self {
+        return Self { edge in
+            guard let lweight = lhs.weight(edge), let rweight = rhs.weight(edge) else { return nil }
+            return lweight * rweight
+        }
+    }
+    
+    static func + (lhs: Self, rhs: Self) -> Self {
+        return Self { edge in
+            if let lweight = lhs.weight(edge) {
+                if let rweight = rhs.weight(edge) { return lweight + rweight }
+                return lweight
+            } else {
+                guard let rweight = rhs.weight(edge) else { return nil }
+                return rweight
+            }
+        }
     }
 }
 
-extension WeightedGraphProtocol {
-    func weightScheme <G> (_ f: @escaping (G.Node) -> Node) -> WeightCarrying<G> where
-        G: WeightedGraphProtocol,
-        G.Weight == Weight
+extension WeightedGraphSchemeProtocol where Self: DirectedGraphSchemeProtocol, Weight: Numeric {
+    
+    static func * <Scheme> (lhs: Self, rhs: Scheme) -> Self where
+    Scheme: WeightedGraphSchemeProtocol,
+    Scheme.Node == Node,
+    Scheme.Weight == Weight
     {
-        return WeightCarrying.build(from: self).pullback(f)
+        return Self { edge in
+            guard
+                let lweight = lhs.weight(edge),
+                let rweight = rhs.weight(from: edge.a, to: edge.b) else { return nil }
+            return lweight * rweight
+        }
+    }
+    
+    static func * <Scheme> (lhs: Self, rhs: Scheme) -> Self where
+        Scheme: UnweightedGraphSchemeProtocol,
+        Scheme.Node == Node
+    {
+        return Self { edge in
+            if let lweight = lhs.weight(edge) {
+                return rhs.contains(from: edge.a, to: edge.b) ? lweight : nil
+            } else {
+                return nil
+            }
+        }
+    }
+    
+    static func * <Scheme> (lhs: Scheme, rhs: Self) -> Self where
+        Scheme: UnweightedGraphSchemeProtocol,
+        Scheme.Node == Node
+    {
+        return rhs * lhs
+    }
+    
+    static func + (lhs: Self, rhs: Self) -> Self {
+        return Self { edge in
```

Comment:
This might be tighter as:

```Swift
switch (lhs.weight(edge), rhs.weight(edge)) {
    case (nil,nil): return nil
    case let (lhs?, rhs?): return lhs + rhs
    case let (lhs, nil): return lhs
    case let (nil, rhs): return rhs
}
```

Or:

```Swift
return lhs.weight(edge).flatMap { lhs in rhs.weight(edge).map { rhs in lhs * rhs }
```

This seems like a perfect place for treating `Optional` as a monad, and using custom operators…

Reply:
Cool!

Reply:
At present, I am going with 
```swift
            lhs.weight(edge).flatMap { weight in
                rhs.containsEdge(from: edge.a, to: edge.b) ? weight : nil
            }
```